Call the application OnResume, on Android Resume instead of Restart.

kvpt opened the pull request:
### Description of Change ###
Fix regression introduced by #4707.

The OnResume worked before due to a side effect.
#4707 fixed the previous activity destroy which incorrectly triggered the page change.
Now that the event state are consistant, this revealed that OnResume event on the application was not called during Android Resume but during Android Restart event.
This PR modify when the SendResume is triggered to be consistant with the Android lifecycle.

### Issues Resolved ### 
- Fixes #7400

### API Changes ###
 None

### Platforms Affected ### 
- Android

### Behavioral/Visual Changes ###
The OnResume method is called on Android Resume

None

### Before/After Screenshots ### 
Not applicable

### Testing Procedure ###
[issue7400.zip](https://github.com/xamarin/Xamarin.Forms/files/3600649/issue7400.zip)
The reproduction project has the framework code included. So you can test with and without the modification.

1. Start reproduction project
2. Set a breakpoint on OnResume method in the App class
3. Move app to background
4. Resume app 
5. Breakpoint must be hit

### PR Checklist ###
- [X] Targets the correct branch
- [X] Tests are passing (or failures are unrelated)

jsuarezruiz commented:
I've tried the sample and I can't reproduce your problem @hartez.
![7035](https://user-images.githubusercontent.com/6755973/64859577-e3c7bf80-d62a-11e9-8cf6-d62d2787d05f.gif)

The only changes after your changes in the sample is a rebase from 4.3.0.
Should I get the same problem with spacing?.

kingces95 commented on Xamarin.Forms.Platform.Android/AppCompat/FormsAppCompatActivity.cs:
```diff
@@ -265,9 +265,16 @@ protected override void OnResume()
+			{
+				_application?.SendResume();
+			}
+
```
Can we move this logic into `OnStateChanged`? I'm not entirely sure why we didn't sprinkle the logic in `OnStateChanged` into the overrides like you have but since we've already established the precedent let's keep it (or recast call the logic as you have here). For one thing, it'll ensure the lollipop code below runs before the callback is placed.

kvpt replied:
@jsuarezruiz I will have some time this weekend and see what can be done.

kvpt replied:
It's done.

kvpt commented:
I tried to run the reproduction on 4.0.0 branch and don't reproduce the issue (as #4707 has been merged in 4.1.0 it's logic).
So the 4.1.0 branch is the right branch to target, no need to retarget on 4.0.0.

kingces95 commented:
Thanks for your hard work re-arranging the code per my request. Much appreciated!

I see we will call `OnResume` when the android lifecycle is resuming: `OnRestart` -> `Start`. To do that, is it necessary to not call `OnResume` during restart: `Stop` -> `Restart`? Might someone have taken a dependency on that behavior however odd? If so, should we add the new behavior and not remove the old behavior?

kvpt commented:
The old behavior trigger the OnResume when activity state is not started yet.
And this cause nasty issues with old and new activity race condition and dispose problem which I fixed with #4707 and #6596.

In #4707 I added this code to prevent paused or stopped activity to respond to the change.
https://github.com/xamarin/Xamarin.Forms/blob/c171efa367fd8e84f040df9882dbec79850cfd85/Xamarin.Forms.Platform.Android/AppCompat/FormsAppCompatActivity.cs#L396
This is this change combined with the fact that the lifecycle was wrong for the restart which break that.
So if you want that the old lifecycle work we will have too revert this change, but it will reintroduce some dispose issues on main page change.

Also if we want the old lifecycle to work and the new at the same time, it will likely cause the Resume event to be called twice.
Perhaps we can use a flags or an OnPlatform (like SendAppearingEventOnResume), to allow the developer to choose the behavior of the lifecycle ?

samhouts commented:
@rmarinho Can you please tell me why this build is failing? Thanks!

rmarinho commented:
@samhouts needs rebase, yaml doesn't have dotnet 3.0 install seems a old branch

samhouts commented:
@kvpt Would you mind rebasing to fix the build? Thanks!

kvpt commented:
I was used 4.1.0 branch has the regression is introduced in this version.
But as the target branch has been changed to master, I rebased on master this time.
It's only two lines, so it will be simple to backport, regardless of the branch.

samhouts commented:
@kvpt I wonder if these failures on API 19 Legacy Renderers are legitimate. Would you mind investigating? Thanks!
![image](https://user-images.githubusercontent.com/7827070/69751864-cfad3d80-1104-11ea-83cd-4bf0d55ac6f9.png)

kvpt commented:
@samhouts This code is only executed when the application lifecycle change state, start/restart/resume.
Nothing in these tests can trigger that, so I think it's not related.
I checked in API 19 emulator and doesn't reproduce the issues.
